Sources/Swift/Tools/Batcher/Batcher.swift

```diff
@@ -0,0 +1,160 @@
+    ) {
+        self.config = config
+        self.storage = storage
+        self.dateProvider = dateProvider
+        self.dispatchQueue = dispatchQueue
+    }
+
+    /// Adds an item to the batcher with the given scope.
+    /// - Parameters:
+    ///   - item: The item to add to the batch
+    ///   - scope: The scope to apply to the item (adds attributes, trace ID, etc.)
+    ///
+    /// - Note: Scope application (attribute enrichment) and the `beforeSendItem` callback are executed
+    ///        synchronously on the caller's thread. Only encoding and buffering happen asynchronously
+    ///        on the batcher's serial dispatch queue. If `config.beforeSendItem` returns `nil`,
+    ///        the item is dropped and not added to the batch.
+    func add(_ item: Item, scope: Scope) {
+        var item = item
+        scope.applyToItem(&item, config: config)
+
+        // The before send item closure can be used to drop items by returning nil
+        // In case it is nil, we can stop processing
+        if let beforeSendItem = config.beforeSendItem {
+            // If the before send hook returns nil, the item should be dropped
+            guard let processedItem = beforeSendItem(item) else {
+                return
+            }
+            item = processedItem
+        }
+
+        dispatchQueue.dispatchAsync { [weak self] in
+            self?.encodeAndBuffer(item: item)
+        }
+    }
+
+    /// Captures all currently batched items synchronously and returns the duration of the operation.
+    /// - Returns: The time taken to capture items in seconds
+    ///
+    /// - Note: This method blocks until all items are captured. The batcher's buffer is cleared after capture.
+    ///
+    /// - Important: This method uses `dispatchSync` to synchronously execute on the batcher's serial dispatch queue.
+    ///              **Do not call this method from within the batcher's dispatch queue or from within the
+    ///              `capturedDataCallback` closure**, as this will cause a deadlock. This method should only be
+    ///              called from external threads or queues (e.g., main thread, app lifecycle callbacks).
+    @discardableResult func capture() -> TimeInterval {
+        let startTimeNs = dateProvider.getAbsoluteTime()
+        dispatchQueue.dispatchSync { [weak self] in
+            self?.performCaptureItems()
+        }
+        let endTimeNs = dateProvider.getAbsoluteTime()
+        return TimeInterval(endTimeNs - startTimeNs) / 1_000_000_000.0 // Convert nanoseconds to seconds
+    }
+
+    /// Encodes and buffers an item, triggering a flush if limits are reached.
+    ///
+    /// - Important: Only call this method from the serial dispatch queue to ensure thread safety.
+    /// - Parameter item: The item to encode and add to the buffer
+    private func encodeAndBuffer(item: Item) {
+        do {
+            let encodedItemsWereEmpty = storage.itemsDataSize == 0
```

Comment:
`l`: `encodedItemsWereEmpty` is a bit misleading, IMO. 
```suggestion
            let storageWasEmpty = storage.itemsDataSize == 0
```